Introduce credential timeout to browser

jakewins opened the pull request:
A much less invasive variant of https://github.com/neo4j/neo4j/pull/4827 since this needed to be back-ported to 2.2.

This should go into 2.2, and get forward merged before the 2.3 PR goes in, then I can rebase the 2.3 PR and sort out the conflicts in there.

jakewins committed:
Make Server config visible via JMX

jakewins commented on enterprise/server-enterprise/pom.xml:
```diff
@@ -257,6 +257,26 @@
         </configuration>
       </plugin>
 
+      <plugin>
```
See eg. the kernel module for what this is - this generates asciidoc from the specified Settings files

jakewins commented on enterprise/server-enterprise/src/main/java/org/neo4j/server/enterprise/EnterpriseServerSettings.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Description( "Configure the operating mode of the database - 'SINGLE' for stand-alone operation or 'HA'" +
+                  "for operating as a member in a cluster." )
+    Setting<String> mode = setting( "org.neo4j.server.database.mode", STRING, "SINGLE" );
```
Moved this in here for good measure, since this was the only config option documented under enterprise server config

jakewins commented on enterprise/server-enterprise/src/main/java/org/neo4j/server/enterprise/EnterpriseServerSettings.java:
```diff
@@ -0,0 +1,44 @@
+{
+    @Description( "Configure the Neo4j Browser to time out logged in users after this idle period. " +
+                  "Setting this to 0 indicates no limit." )
+    Setting<Long> browser_credentialTimeout = setting("dbms.browser.credentialTimeout", DURATION, "0");
```
These will get picked up by the Browser in a separate PR and used to trigger browser behavior

davidegrohmann replied:
Wonder if it makes sense to have `BrowserSettings` for browser.

davidegrohmann commented on community/server/src/main/java/org/neo4j/server/web/ServerInternalSettings.java:
```diff
@@ -83,6 +83,7 @@
             PATH, File.separator + "etc" + File.separator + "neo" + File.separator + ServerInternalSettings.DB_TUNING_CONFIG_FILE_NAME );
 
+    // TODO: Remove this in 2.3 with the rest of the config refactorings there, this now lives in EnterpriseServerSettings
     public static final Setting<String> legacy_db_mode = setting( "org.neo4j.server.database.mode", STRING, "SINGLE" );
```
Why can't we remove this now?

davidegrohmann commented on community/jmx/src/main/java/org/neo4j/jmx/Kernel.java:
```diff
@@ -27,7 +27,7 @@
 public interface Kernel
 {
-    final String NAME = "Kernel";
+    String NAME = "Kernel";
```
I don't understand this change.

jotomo replied:
The _final_ keyword is redundant in an interface. So this is likely IntelliJ's suggested quickfix applied :-)

davidegrohmann commented:
@jakewins it looks good to me

oskarhane committed:
Introduce security policies in the browser.

- Follow server config policies for storing credentials and credential timeout on enterprise.
- Disconnect user on hit of credentials timeout.
- `:config storeCredentials:true/false` toggles credentials in local storage in browser.
- Credentials will be stored if `server.storeCredentials AND client.storeCredentials`.
- Display policies on `:server connect` and `:server status` frames
  when connected.
